flink-shaded-hadoop-2-parent/pom.xml

```diff
@@ -0,0 +1,179 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>org.apache.flink</groupId>
+        <artifactId>flink-shaded</artifactId>
+        <version>10.0</version>
+        <relativePath>..</relativePath>
+    </parent>
+
+    <artifactId>flink-shaded-hadoop-2-parent</artifactId>
+    <name>flink-shaded-hadoop-2-parent</name>
+    <packaging>pom</packaging>
+    <version>10.0</version>
+
+    <properties>
+        <!-- never use this in the parent module; otherwise it would need to be versioned by hadoop version; this must
```

Comment:
If I understand correctly, `hadoop.version` must not be used inside this pom. After reading the comment, I do not understand why. Can you elaborate on this?

Reply:
It's similar to what we have to do with Scala; if you have a scala dependency then you need a scala suffix in the name, to differentiate the resulting artifacts for different scala versions.

The short answer is that you can introduce version mismatches in the dependency set, which could mean that  if you depend on `flink-shaded-hadoop-2` 2.4.1, then the parent pom that it references may contain hadoop 2.9.X dependencies.

Here's the long answer:

If you use this property in the pom then the `pom` we release for `flink-shaded-hadoop-2-parent` contains a dependency/something with a specific hadoop version, and the pom is hence parameterized based on what `hadoop.version` is set.

That's not necessarily a problem; if the version of the parent is parameterized as well (i.e., `<version>${hadoop.version}-10.0</version>`) then things would work, however that cannot be done since the parent reference in child modules cannot have parameters.
This is a bit subtle but obvious once you realize it; if the parent is parameterized, and the value of that parameter resides in the parent, then we can never resolve the parent since to do so we would already have to know who the parent is, as otherwise we can't retrieve the parameter value. To fix this you'd have to copy the `hadoop.version` property into every related module, which is obviously terrible.

So why would it be a problem otherwise?
For one it's no longer well-defined what you are actually releasing. During the release process we create several _different_ poms for the parent module, which are identified by maven as one and the same. You don't really know which one will actually be released.

This _also_ means that the _actual version_ of the pom that the child modules were built against _is not actually being released_, which is where we're coming back to the short answer. Since _some_ version of the parent is being released, which _all_ child modules reference, there will always be a version mismatch somewhere between the hadoop dependencies (except for one version).

Reply:
This is _somewhat_ fine for the `uber` module since it bundles everything anyway and the effective pom contains no dependencies, but the non-uber module may be affected by this.